### PR TITLE
Trigger manual cleanups in WAVM 

### PR DESCRIPTION
**Pull request by axic**

### CODE REVIEW

**Comment by axic** [comment]
This doesn't eliminates all leaks, but reduces them significantly.

File: src/wavm.cpp [hunk 1]
```diff
@@ -199,10 +199,18 @@ ExecutionResult WavmEngine::execute(
...
+    ExecutionResult result = internalExecute(context, code, state_code, msg, meterInterfaceGas);
+    // And clean up mess left by this run.
+    Runtime::collectGarbage();
+    return result;
+  } catch (exception const& e) {
+    // And clean up mess left by this run.
+    Runtime::collectGarbage();
+    // We only catch this exception here in order to clean up garbage..
+    // TODO: hopefully WAVM is fixed so that this isn't needed
+    throw e;
```
Comment (axic) [inline src/wavm.cpp]: Should be `throw;` to not loose derived exceptions.

**Comment by axic** [comment]
Do not review/merge yet - have a lot more fixes.

File: src/wavm.cpp [hunk 2]
```diff
@@ -268,21 +275,28 @@ ExecutionResult WavmEngine::internalExecute(
...
-    },
-    [&](Runtime::Exception&& exception) {
-      // FIXME: decide if each of the exception fit into VMTrap/InternalError
-      ensureCondition(false, VMTrap, Runtime::describeException(exception));
-    }
-  );
+    );
+  } catch (exception const& e) {
+    // clean up
+    wavm_host_module::interface.pop();
```
Comment (chfast) [inline src/wavm.cpp]: The `interface` contains raw pointers, are those deleted?

Comment (chfast) [reply src/wavm.cpp]: Is this `wavm_host_module` required by WAVM? It's not thread-save to have a global structure handling execution.

File: src/wavm.cpp [hunk 3]
```diff
@@ -268,21 +275,28 @@ ExecutionResult WavmEngine::internalExecute(
...
-    [&] {
-      try {
-        vector<IR::Value> invokeArgs;
-        Runtime::invokeFunctionChecked(wavm_context, mainFunction, invokeArgs);
-      } catch (EndExecution const&) {
-        // This exception is ignored here because we consider it to be a success.
-        // It is only a clutch for POSIX style exit()
+  // TODO: this is terrible, need to clean up the handling of `interface`
+  try {
+    Runtime::catchRuntimeExceptions(
```
Comment (chfast) [inline src/wavm.cpp]: Does this actually execute anything?

File: src/wavm.cpp (hunk 2 above)
Comment (axic) [reply src/wavm.cpp]: It is required due to that fact the host functions can be declared via macros only :(

File: src/wavm.cpp (hunk 3 above)
Comment (axic) [reply src/wavm.cpp]: Yeah, it catches wasm runtime errors (but passes through exceptions in host functions).

File: src/wavm.cpp (hunk 2 above)
Comment (axic) [reply src/wavm.cpp]: > The `interface` contains raw pointers, are those deleted?

That code was I think suggested by you, can you please check?

Comment (chfast) [reply src/wavm.cpp]: Yes, it is allocated on the stack.

File: src/wavm.cpp [hunk 4]
```diff
@@ -271,21 +279,28 @@ ExecutionResult WavmEngine::internalExecute(
...
-    },
-    [&](Runtime::Exception&& exception) {
-      // FIXME: decide if each of the exception fit into VMTrap/InternalError
-      ensureCondition(false, VMTrap, Runtime::describeException(exception));
-    }
-  );
+    );
+  } catch (exception const&) {
+    // clean up
+    wavm_host_module::interface.pop();
```
Comment (axic) [inline src/wavm.cpp]: This is actually fixing the problem of exceptions not cleaning up between calls.

**Pull request merged**